### DISCUSSION TITLE
Route parameter not replacing state on browser url when updated

### DISCUSSION THREAD

**Discussion by jDiazSenn** [livewire/livewire#6939] [7. Bugs (archived)]
### Livewire version

v3.0.5

### Which PHP version are you using?

PHP 8.2

### Steps To Reproduce

1.  Upgraded Livewire from v2 to v3
2. Create new component `sail artisan make:livewire TestComponent`
3. Declare $testParam variable and a method to change the value of it. 
4. Create route to access the component `Route::get('/test-component/{testParam?}', App\Livewire\TestComponent::class);`
5. On the component view, simply add a link with a wire:click that calls changeParam function and updates the value of it.
6. On v2 the behaviour replace the url with the new param value, where on v3 the param updates its value but no change on the url.

**Livewire Component**
```
<?php

namespace App\Livewire;

use Livewire\Component;

class TestComponent extends Component
{
    public string $param = '';

    public function changeParam(string $param): void
    {
        $this->param = $param;
    }


    public function render(): \Illuminate\Contracts\View\View|\Illuminate\Foundation\Application|\Illuminate\Contracts\View\Factory|\Illuminate\Contracts\Foundation\Application
    {
        return view('livewire.test-component');
    }
}
```

**Component View** 'livewire.test-component'

```
<div>
    <a href="javascript:void(0);" wire:click="changeParam('something')">Change param</a>
</div>
```

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by arthurshlain** [comment]
Hi. Add `#[Url]` attribute to param:

```php
<?php

namespace App\Livewire;

use Livewire\Attributes\Url;
use Livewire\Component;

class TestComponent extends Component
{
    #[Url]
    public string $param = '';

    public function changeParam(string $param): void
    {
        $this->param = $param;
    }

    public function render()
    {
        return view('livewire.test-component');
    }
}
```

If you with to rename query string param: 

```php
#[Url(as: 'q')]
public string $param = '';
```

Route can be simplified now:

```php
Route::get('/test-component', App\Livewire\TestComponent::class);
```

https://livewire.laravel.com/docs/url

**Reply by jDiazSenn** [reply]
Hi, thanks for the reply.

I am aware of the Url attribute, that will put the data as a query parameter where my goal is to put it as a route parameter.

I.e:
`Route::get('/test-component/{param?}',  App\Livewire\TestComponent::class);`

Which would produce
`https://url.com/test-component/text`
Instead of:
`https://url.com/test-component?param=text`

On version 2, when a route parameter was changed from the controller (on my example with the changeParam function), it was replacing it on the url, allowing me to navigate back inside the same livewire component.

Again, thanks for the quick reply.

**Reply by arthurshlain** [reply]
Hm, really, it's not so easy to do from Livewire 3. I can do it this way with named route:

`Route::get('/test-component/{param?}', App\Livewire\TestComponent::class)->name('test-route');`

```php
class TestComponent extends Component
{
    public string $param = '';
    public function changeParam(string $param): string
    {
        $this->param = $param;
        return route('test-route', ['param' => $param]);
    }
}

```

```blade
<div>
    <p>{{ $param }}</p>
    <a x-data href="javascript:void(0);"
       wire:click="changeParam('something').then(function(url){ history.replaceState(null, null, url) })">
        Change param
    </a>
</div>
```

I use something like this to update URL's in my project when modals are open.

**Comment by drifteaur** [comment]
This is a regression from Livewire 2, also related: #7253

**Comment by djeknet** [comment]
have same problem!

**Comment by thorstenEURESA** [comment]
have same problem.

**Comment by calebporzio** [comment]
Yes, this behavior has been removed in V3. V2 manned the URL on the server which caused lots of problems. V3 manages the URL on the client and only manipulates the query string.

wire:navigate should help fill this gap in V3 for some people as well.
